pkg/controllers/provisioning/v1alpha1/reallocation/suite_test.go

```diff
@@ -0,0 +1,190 @@
+
+			ExpectCreated(ns.Client, p)
+			ExpectEventuallyReconciled(ns.Client, p)
+
+			Expect(ns.Client.Get(context.Background(), client.ObjectKey{Name: node.Name}, node)).To(Succeed())
+			Expect(node.Labels).To(HaveKeyWithValue(v1alpha1.ProvisionerPhaseLabel, v1alpha1.ProvisionerUnderutilizedPhase))
+			Expect(node.Annotations).To(HaveKey(v1alpha1.ProvisionerTTLKey))
+		})
+
+		It("should remove labels from utilized nodes", func() {
+			p := &v1alpha1.Provisioner{
+				ObjectMeta: metav1.ObjectMeta{Name: strings.ToLower(randomdata.SillyName()), Namespace: ns.Name},
+				Spec: v1alpha1.ProvisionerSpec{
+					Cluster: &v1alpha1.ClusterSpec{Name: "test-cluster", Endpoint: "http://test-cluster", CABundle: "dGVzdC1jbHVzdGVyCg=="},
+				},
+			}
+			node := test.NodeWith(test.NodeOptions{
+				Labels: map[string]string{
+					v1alpha1.ProvisionerNameLabelKey:      p.Name,
+					v1alpha1.ProvisionerNamespaceLabelKey: p.Namespace,
+					v1alpha1.ProvisionerPhaseLabel:        v1alpha1.ProvisionerUnderutilizedPhase,
+				},
+				Annotations: map[string]string{
+					v1alpha1.ProvisionerTTLKey: time.Now().Add(time.Duration(100) * time.Second).Format(time.RFC3339),
+				},
+			})
+			pod := test.PodWith(test.PodOptions{
+				Name:       strings.ToLower(randomdata.SillyName()),
+				Namespace:  p.Namespace,
+				NodeName:   &node.Name,
+				Conditions: []v1.PodCondition{{Type: v1.PodReady, Status: v1.ConditionTrue}},
+			})
+
+			ExpectCreatedWithStatus(ns.Client, node)
+			ExpectCreatedWithStatus(ns.Client, pod)
+
+			ExpectCreated(ns.Client, p)
+			ExpectEventuallyReconciled(ns.Client, p)
+
+			nodeCheck := &v1.Node{}
+			Expect(ns.Client.Get(context.Background(), client.ObjectKey{Name: node.Name}, nodeCheck)).To(Succeed())
+			Expect(nodeCheck.Labels).ToNot(HaveKey(v1alpha1.ProvisionerPhaseLabel))
+			Expect(nodeCheck.Annotations).ToNot(HaveKey(v1alpha1.ProvisionerTTLKey))
+		})
+
+		It("should terminate nodes with expired TTL", func() {
+			p := &v1alpha1.Provisioner{
+				ObjectMeta: metav1.ObjectMeta{Name: strings.ToLower(randomdata.SillyName()), Namespace: ns.Name},
+				Spec: v1alpha1.ProvisionerSpec{
+					Cluster: &v1alpha1.ClusterSpec{Name: "test-cluster", Endpoint: "http://test-cluster", CABundle: "dGVzdC1jbHVzdGVyCg=="},
+				},
+			}
+			node := test.NodeWith(test.NodeOptions{
+				Labels: map[string]string{
+					v1alpha1.ProvisionerNameLabelKey:      p.Name,
+					v1alpha1.ProvisionerNamespaceLabelKey: p.Namespace,
+					v1alpha1.ProvisionerPhaseLabel:        v1alpha1.ProvisionerUnderutilizedPhase,
+				},
+				Annotations: map[string]string{
+					v1alpha1.ProvisionerTTLKey: time.Now().Add(time.Duration(-100) * time.Second).Format(time.RFC3339),
+				},
+			})
+			ExpectCreatedWithStatus(ns.Client, node)
+
+			ExpectCreated(ns.Client, p)
+			ExpectEventuallyReconciled(ns.Client, p)
+
+			nodeCheck := &v1.Node{}
+			Expect(ns.Client.Get(context.Background(), client.ObjectKey{Name: node.Name}, nodeCheck)).ToNot(Succeed())
```

Comment:
This may fail in a spurious way (like a network connectivity issue). I wonder if we could write a check for does not exist? expectations.ExpectToNotExist(client, object). There's a good example of the NotFound error checking in the genericcontroller

Reply:
I will use this. Thanks.